### DISCUSSION TITLE
problem publishing large file to azure blob storage

### DISCUSSION THREAD

**Discussion by kobelavaerts** [nextflow-io/nextflow#2610] [Q&A]
Hello everyone,

I hope it's OK for me to ask this here.
I have run into a problem when trying to publish a large file to azure blob storage.

We run our pipeline with executor Local but want to write some of our output files together with the process `.command.*` files and the report, timeline and trace files to an azure blob storage container.

This works for the most part except for one file.
We have one process that creates a compressed `tar.gz` archive of intermediate output files. This is a relatively big file of 1.7 TB. When trying to publish this file to the blob storage, nextflow trows `error 409. The uncommitted block count cannot exceed the maximum limit of 100,000 blocks.`

`.nextflow.log` excerpt:

```
Jan-27 19:01:49.779 [FileTransfer-thread-67] ERROR c.a.s.common.StorageOutputStream - com.azure.storage.blob.models.B
lobStorageException: Status code 409, "<U+FEFF><?xml version="1.0" encoding="utf-8"?><Error><Code>BlockCountExceedsLi
mit</Code><Message>The uncommitted block count cannot exceed the maximum limit of 100,000 blocks.
RequestId:1dbd201e-a01e-0035-6db0-132d0d000000
Time:2022-01-27T19:01:49.7131543Z</Message></Error>"
```
All other output files and log files of other processes are successfully published to the azure blob container.

We thought maybe the file was to big, but when we try to upload the file manually to the same azure blob container using `azcopy` we don't get this error and the file is successfully uploaded.

As I'm not able to resolve the problem I would appreciate any suggestions for fixing this issue.

more information below:
```
Nextflow version 21.10.0, build 5640 (11-11-2021 18:27 UTC)
```

`nextflow.config` file excerpt
```

azure {
  storage {
    accountName = params.azstorage
    sasToken = params.azsastoken
  }
}

// specify the output folders
params.tostoredir = "az://" + params.azblob + "/ToStore"
params.tosenddir = "az://" + params.azblob + "/ToSend"
params.logdir = params.tostoredir + "/log"

// Save the run report and trace files in azure blob storage
report {
  enabled = true
  file = "az://${params.azblob}/ToStore/log/${params.runname}-report.html"
}
trace {
  enabled = true
  file = "az://${params.azblob}/ToStore/log/${params.runname}-trace.txt"
}
timeline {
  enabled = true
  file = "az://${params.azblob}/ToStore/log/${params.runname}-timeline.html"
}
```

nextflow process:
```
process create_archive{
    publishDir "${params.tosenddir}", pattern: "*.{gz,txt}", mode: "copy"
    publishDir "${params.logdir}/${task.process}/${task.hash}", pattern: ".*", mode: "copy"

    errorStrategy "retry"

    input:
    val(project_name)
    val(demuxdone)

    output:
    tuple val(project_name), file("**")
    path(".*")

    script:
    """
    tar --use-compress-program="pigz -p ${task.cpus}" -cf ${params.runname}-RawData-${project_name}.tar.gz -C ${params.rawdata} ${project_name}
    md5sum ${params.runname}-RawData-${project_name}.tar.gz > ${params.runname}-RawData-${project_name}.md5sum.txt
    """
}
```
`.command.sh`:
```
#!/bin/bash -ue
tar --use-compress-program="pigz -p 12" -cf 220124_A01044_0162_BH5VJYDSX3-RawData-3971.tar.gz -C /mnt/sdb1/220124_A01044_0162_BH5VJYDSX3/RawData 3971
md5sum 220124_A01044_0162_BH5VJYDSX3-RawData-3971.tar.gz > 220124_A01044_0162_BH5VJYDSX3-RawData-3971.md5sum.txt
```
Please let me know if you want additional information.

**Comment by kobelavaerts** [comment]
I did some more digging and have a bit of additional information about the issue.

We did some calls to the azure [Blob Service REST API](https://docs.microsoft.com/en-us/rest/api/storageservices/get-block-list#:~:text=Uncommitted%20Block%20List%3A%20The%20list,form%20part%20of%20the%20blob.) to look at the uncommitted and committed blocks of the file.

When listing the blocks from the blob where the `tar.gz` file should have been written to, we see that there are no committed blocks and there are indeed `100 000` uncommitted blocks. We also see that each block is `4194304 B`, corresponding to `4.0 MB`. If we have 100k blocks of this size, we get `390 GB`. This explains why the `tar.gz` file failed to upload since `1.7 TB` is more than the maximum allowed `390 GB`.

example command used:
```
curl \
'https://[my-account].blob.core.windows.net/[blob-container]/[path/to/blob/tar.gz]?comp=blocklist&blocklisttype=all&[SAS-token] \
-o azure_blocks.txt \
-H "Accept: application/xml"
```

When we use the same commands on a successfully uploaded blob, we see only committed blocks and no uncommitted blocks. The committed blocks here are again `4.0 MB`.

When we look at the blob of the same `tar.gz` file that we uploaded using `azcopy` ourselves, we see that there are `27746` committed blocks with a size of `67108864 B` or `64 MB` per block . In total this corresponds to `1.69 TB` which is the correct size of the file. 

When going through the source code of the [nf-azure plugin](https://github.com/nextflow-io/nextflow/tree/master/plugins/nf-azure) it seems like nextflow uses `azcopy` as well to transfer files to and from blob containers. [link](https://github.com/nextflow-io/nextflow/blob/master/plugins/nf-azure/src/main/nextflow/cloud/azure/file/AzBashLib.groovy). The plugin seems to use `azcopy version 10.13.0` which is the same version we have installed.

My question now is why is the block size limited to `4 MB`  and can the block size be increased to allow us to publish larger files to our blob container?
@pditommaso would you be able to help with this?

**Reply by abhi18av** [reply]
Hi @kobelavaerts ,

It seems that `--block-size-mb` option might be of help here.

```
$ azcopy copy --help | grep 'size'
      --block-size-mb float                  Use this block size (specified in MiB) when uploading to Azure Storage, and downloading from Azure Storage. The default value is automatically calculated based on file size. Decimal fractions are allowed (For example: 0.25).

```


Have you tried this option in isolation (not in context of Nextflow)?

If so, what's the optimal value which helped your push through this file to the blob container?

**Reply by kobelavaerts** [reply]
Hi @abhi18av ,

Thanks for the reply.

We indeed knew about the `--block-size-mb` option and have tried it in isolation. This option works and lets you specify the block size. We don't know the optimal value but we know it works when specifying `--block-size-mb 64`. I assume it works anywhere from `35.7MB` per block until the maximum of `400MB` per block [link](https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs), taking into account the file size (1.7TB) and the maximum amount of committed blocks (50 000). Nextflow constructs the `azcopy` command however, over which we have no control. Meaning that it seems we can't specify the block size for the `azcopy` command used by Nextflow.

When we use `azcopy` manually however (outside of Nextflow), without the `--block-size-mb` option, the block size seems to change automatically based on the total file size like I mentioned in the original post. 
When Nextflow uses `azcopy`, the block size doesn't seem to change based on the file size automatically. I don't understand why this is the case. Maybe the `--block-size-mb` is restricted somewhere in the source code of the `nf-azure` plugin but I couldn't find it.

To solve this problem, I think we should either be able to specify the block size used by nextflow, or the `azcopy` command used by Nextflow should change the block size based on the total file size.

**Reply by abhi18av** [reply]
>  I don't understand why this is the case. Maybe the --block-size-mb is restricted somewhere in the source code of the nf-azure plugin but I couldn't find it.

This is an interesting point, I also did a quick [search on the code](https://github.com/nextflow-io/nextflow/search?q=--block-size-mb&type=code) and didn't find any such hard limit therefore I'm not sure why the `azcopy` tool which used in default mode shouldn't accommodate as per the file size.

But in any case, I think that some options could be added into the `AzBashLib` file similar to what is available for [aws.client](https://www.nextflow.io/docs/edge/config.html#scope-aws) to allow an explicit setting for this option.

Please allow me some time to look into this, will share my updates here.

**Reply by abhi18av** [reply]
Hi @kobelavaerts , 

Regarding the accommodation of various `azcopy` options, I have submitted a PR here https://github.com/nextflow-io/nextflow/pull/2632 using which it's possible to customize the `--block-blob-tier` and `--block-size-mb` options

**Reply by kobelavaerts** [reply]
Hi @abhi18av,

Thanks a lot for looking into the issue and working on a solution!
Do you have an estimate on when the pull request will be approved so we can test out the options?

**Reply by abhi18av** [reply]
Hi @kobelavaerts ,

Happy to be of help!

About the merge, I'm not sure. In the meantime, you can test the code in the branch using a compiled copy of Nextflow with the instructions [here](https://github.com/nextflow-io/nextflow#build-from-source) and then using the `/path/to/nextflow/source/launch.sh run ...` with the custom options.

Hopefully, this will be a part of the next edge release. Your feedback can help in finalizing the design 👍

**Reply by Tomekke** [reply]
Hello @abhi18av 

I'm a colleague of @kobelavaerts. First, thank you for your help on this issue. We tried out the solution, but I'm afraid it didn't work as wished. The main code blocks will be posted after our summary.
We ran this on a naked docker container, so we know we have no influence from other packages. We started by doing `git clone https://github.com/abhi18av/nextflow.git -b abhinav/azcopy-config` to get your code. Once in the nextflow folder we did a `make compile`. This was successful and we launched a small .nf script via `/opt/nextflow/launch mainf.nf`. This worked, sent the files to our blob storage. Afterwards we checked with curl, but still gave us a block size back of 4MB:
```sh
curl '[ACCOUNT_NAME]/ToStore/tasklib.tar.gz?comp=blocklist&blocklisttype=all&[SAS-TOKEN] -o az_output_tar.txt -H "Accept: application/xml"
```
```xml
<?xml version="1.0" encoding="utf-8"?>
<BlockList>
    <CommittedBlocks>
        <Block>
            <Name>ZjNiMzJkODItMDkxNC00YTI3LThmYzAtNzkyYzIwZjkwNTg0</Name>
            <Size>4194304</Size>
        </Block>
        ...
   </CommittedBlocks>
   <UncommittedBlocks />
</BlockList>
```

We downloaded and tried manually azcopy, which did gave the specified block size
```sh
./azcopy cp /mnt/input/tasklib.tar.gz '[ACCOUNT_NAME]/ToStore/?[SAS-TOKEN]' --block-size-mb "64" --block-blob-tier "Hot"
```
```xml
<?xml version="1.0" encoding="utf-8"?>
<BlockList>
    <CommittedBlocks>
        <Block>
            <Name>MzllMDlhNTMtNzIzMC1mNzQ3LTcwZDktNzJhOGI0ODA0YWJk</Name>
            <Size>67108864</Size>
        </Block>
        ...
   </CommittedBlocks>
   <UncommittedBlocks />
</BlockList>
```
We tried searching in the code for the log file you specify in issue #2632 , `.../.azcopy_log`, but we couldn't seem to find it. In an act of desperation I did some batch sed and changed in our downloaded local nextcloud folder all azcopy cp occurrences (so especially those from AzBashLib.groovy) from `$AZCOPY_BLOCK_SIZE_MB` to `"64`, with a `make compile` behind, but without avail. The block size stays 4MB. I tried to find a way to log the specified azcopy command when using launch.sh, but haven't found a way yet. 

Anyway, hope this complete enough for you to help us further. If you need anything more please let us know.
Kind regards
Tomekke

_nextflow.config_
```groovy
params {
    azstorage = "[ACCOUNT_NAME]"
    azsastoken = '[SAS-TOKEN]'
    azblob = "bloboutput"
    tarfile = "/mnt/input/tasklib.tar.gz"
}

azure {
      storage {
        accountName = params.azstorage
        sasToken = params.azsastoken
      }
      azcopy {
        blockSize = "64"
        blobTier = "Hot"
      }
    }

    // specify the output folders
    params.tostoredir = "az://" + params.azblob + "/ToStore"
    params.tosenddir = "az://" + params.azblob + "/ToSend"
```

_main.nf_
```nextflow
#!/usr/bin/env nextflow

// Activate  the 2
nextflow.enable.dsl=2

process hoofd{
    publishDir params.tostoredir, mode: 'copy'

    input:
    path(tarfile)

    output:
    path("*", includeInputs: true)

    script:
    """
    ls -l > ls_output.txt
    """
}

workflow{
    inputFile = Channel.fromPath(params.tarfile)
    hoofd(inputFile)
}
```

**Reply by abhi18av** [reply]
Happy to help and investigate this further @Tomekke, also thanks for further clarification 👍 

By any chance, if the task node is still alive could you please check in the `shared` folder (maybe?) if the `.azcopy_log` logs are there? I think that someone else might be at play here and would like to confirm that at least the options are being used by the `azcopy` command as expected.

I'll give it a look over the weekend but I'm pinging @vsmalladi in case he has some insights to offer here.

**Reply by Tomekke** [reply]
We will check it out if we can find the log file. Just to be clear, we are not working with Azure Batch Services (yet). We are running Nextflow locally on an Azure VM and push our final data to a blob storage (so only way is via azcopy).
Cheers

**Reply by vsmalladi** [reply]
@Tomekke so if i understand you correctly you are doing the following:

1. Get Azure VM
2. Launch nextflow locally on Azure VM
3. AZ copy files from Azure VM to blob storage

Another way is to use Azure File share to mount to VM and just deposit all the files in the storage account.

https://docs.microsoft.com/en-us/azure/storage/files/storage-how-to-use-files-linux?tabs=smb311

What are you thoughts on that @abhi18av

**Reply by Tomekke** [reply]
@vsmalladi You're correct. We go from Azure VM -> Blob storage but then we bring back our data to the local server. And due to security reasons outside of our control we can't mount a file share on that local server. So in short we need azcopy to copy from Azure VM to blob storage, hence our request.
Cheers
Edit: Replying directly to thread.

**Reply by abhi18av** [reply]
> Another way is to use Azure File share to mount to VM and just deposit all the files in the storage account.

Agreed, for transferring large files generated within a task, to blob storage `azcopy` might not be the optimal choice - instead you can make use of [Azure File Share integration](https://www.nextflow.io/docs/latest/azure.html#azure-file-shares) with Nextflow which should alleviate these transfer pains.

However, given the permission constraint mentioned above, I think that [blob fuse](https://github.com/Azure/azure-storage-fuse) might be worth exploring since it can be mounted on the VM and offers further customizations and caching.

That being said, I'm still looking into why the block size parameter isn't as expected. Apologies for the delay in update here, got caught up in something.

**Reply by kobelavaerts** [reply]
@vsmalladi Thank you for the suggestion, like @Tomekke mentioned this is unfortunately not an option for us.

@abhi18av Thanks a lot for your efforts, I'm happy you're still looking into it.

Indeed as my colleague mentioned, we run the pipeline not with the `azurebatch` executor but with `local` executor. My apologies for the misunderstanding, I did mention it in my original post but I could have been clearer.

We didn't find the `.azcopy_log` file but when looking at the source code I think it might have something to do with us not using the `azurebatch` executor. Sadly this means we can't look at the `azcopy` command used or the output.

For me it's not clear which classes and methods are only loaded when using `azurebatch` executor vs when the `azure` scope is used without the `azurebatch` executor. But if there are differences there, I was thinking if it's maybe possible that an older version of `azcopy` is used when the `azurebatch` executor isn't set? The reason I bring this up is because from the [documentation](https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs) it seems that with older versions of `azcopy` (service versions prior to 2016-05-31), the block-size limit was `4MB` which would correspond with the block size of `4MB` we see on all our blobs published by nextflow.

I'm a little out of my depth here but what do you think about it? Maybe it's worth checking out.

**Reply by abhi18av** [reply]
Thanks for the link @kobelavaerts , using `--log-level debug` (the default is `info`) - I was able to confirm that the API version used is `2019-12-12` 

```
   Content-Length: [10485760]
   User-Agent: [AzCopy/10.13.0 Azure-Storage/0.14 (go1.17.2; darwin)]
   X-Ms-Client-Request-Id: [xxxxxxxxxxxxxxxxxxxxxxxxxxx]
   X-Ms-Version: [2019-12-12]

```

Given that you are also using the same version of `azcopy` as per https://github.com/nextflow-io/nextflow/discussions/2610#discussioncomment-2112362 . Therefore, I am inclined to think that might not be the reason why this is happening.

In any case, I have been able to at least reproduce this issue on my end and am currently working on enhancements to the PR to address this issue.

**Reply by abhi18av** [reply]
**Updates:**


How I reproduced it locally.  Basically, have a process which produces a file of a specific size 

```nextflow

process PUBLISH_REPORTS {
    publishDir "${params.outdir}"

    output:
        path("*.tmp")

    script:
        """
        mkfile -n 50m file_50mb.tmp
        """

}


```

- Setup the creds in `nextflow.config` for the following 

```nextflow

profiles {

    azurebatch {

        params {
            outdir = "az://${AZURE_STORAGE_CONTAINER_NAME}/results"
        }

        azure {
            azcopy {
                blockSize = "10"
                blobTier = "Hot"
            }

            storage {
                accountName = "$AZURE_STORAGE_ACCOUNT_NAME"
                accountKey = "$AZURE_STORAGE_ACCOUNT_KEY"
            }
        }
    }
}

```
 
- Then using the compiled `abhinav/azcopy-config` branch (mentioned in the thread above) you can launch the pipeline 

```
$ launch.sh run main.nf -profile azurebatch

```


> ... we run the pipeline not with the azurebatch executor but with local executor.

So, the crucial bit here is that when the Azure Batch executor isn't used, which is the only place where [`AzBashLib`](https://github.com/nextflow-io/nextflow/blob/774301bf53ed80cb9da6a99151c7dd30338f106b/plugins/nf-azure/src/main/nextflow/cloud/azure/batch/AzFileCopyStrategy.groovy#L83) is used and consequently, the `AzCopyOpts` isn't being invoked at all.

Instead, the Java SDK for Azure Storage is used for transferring the file to the blob container using [this code here](https://github.com/nextflow-io/nextflow/blob/master/plugins/nf-azure/src/main/nextflow/cloud/azure/nio/AzFileSystem.groovy#L410-L425).

**TLDR:** The `azcopy` isn't used to transfer files from local executor to the blob storage. However, if the executor is `azurebatch` then `azcopy` will be invoked.

Given that there is a `BLOB_DEFAULT_UPLOAD_BLOCK_SIZE` field [within the `BlobClient` class](https://docs.microsoft.com/en-us/java/api/com.azure.storage.blob.blobclient?view=azure-java-stable#fields), I think those are the options we need to look into, in order to accomplish what you're requesting wrt the `local` executor and publishing files to `Blob container`.

This _might_ be a major refactor for the `copy strategy` to allow further customizability, however, that's beyond the time I have on my hands atm 😅 

I do suggest that you look into the `blobfuse` for alleviating this need in the meantime and mount the blob container in your VM.

**Comment by abhi18av** [comment]
(Reposting the response in the thread here to summarize the overall issue.)


**Updates:**


How I reproduced it locally.  Basically, have a process which produces a file of a specific size 

```nextflow

process PUBLISH_REPORTS {
    publishDir "${params.outdir}"

    output:
        path("*.tmp")

    script:
        """
        mkfile -n 50m file_50mb.tmp
        """

}


```

- Setup the creds in `nextflow.config` for the following 

```nextflow

profiles {

    azurebatch {

        params {
            outdir = "az://${AZURE_STORAGE_CONTAINER_NAME}/results"
        }

        azure {
            azcopy {
                blockSize = "10"
                blobTier = "Hot"
            }

            storage {
                accountName = "$AZURE_STORAGE_ACCOUNT_NAME"
                accountKey = "$AZURE_STORAGE_ACCOUNT_KEY"
            }
        }
    }
}

```
 
- Then using the compiled `abhinav/azcopy-config` branch (mentioned in the thread above) you can launch the pipeline 

```
$ launch.sh run main.nf -profile azurebatch

```


> ... we run the pipeline not with the azurebatch executor but with local executor.

So, the crucial bit here is that when the Azure Batch executor isn't used, which is the only place where [`AzBashLib`](https://github.com/nextflow-io/nextflow/blob/774301bf53ed80cb9da6a99151c7dd30338f106b/plugins/nf-azure/src/main/nextflow/cloud/azure/batch/AzFileCopyStrategy.groovy#L83) is used and consequently, the `AzCopyOpts` isn't being invoked at all.

Instead, the Java SDK for Azure Storage is used for transferring the file to the blob container using [this code here](https://github.com/nextflow-io/nextflow/blob/master/plugins/nf-azure/src/main/nextflow/cloud/azure/nio/AzFileSystem.groovy#L410-L425).

**TLDR:** The `azcopy` isn't used to transfer files from local executor to the blob storage. However, if the executor is `azurebatch` then `azcopy` will be invoked.

Given that there is a `BLOB_DEFAULT_UPLOAD_BLOCK_SIZE` field [within the `BlobClient` class](https://docs.microsoft.com/en-us/java/api/com.azure.storage.blob.blobclient?view=azure-java-stable#fields), I think those are the options we need to look into, in order to accomplish what you're requesting wrt the `local` executor and publishing files to `Blob container`.

This _might_ be a major refactor for the `copy strategy` to allow further customizability, however, that's beyond the time I have on my hands atm 😅 

I do suggest that you look into the `blobfuse` for alleviating this need in the meantime and mount the blob container in your VM.
